### PR TITLE
Test muxer: "happy path"

### PR DESCRIPTION
**Pull request by dvdplm**
Taking inspiration from the `two_peers` tests for `mplex`, adds a new `test_muxer` crate that exports a `test_muxer!` macro. Using the macro implementers of `StreamMuxer` can test themselves.

This PR is a start for a more extensive test suite and currently only tests for rather simple situations using the default configs for `yamux` and `mplex`. Notably it does not include any tests for error conditions as these have proved to be difficult to generate reliably for each muxer implementation.

Some future directions for this work would be:
- test different configuration options
- add more "happy path" test cases
- add more tests cases for error conditions and edge cases

### CODE REVIEW

File: muxers/mplex/src/lib.rs [hunk 1]
```diff
@@ -583,10 +588,23 @@ enum OutboundSubstreamState {
...
     current_data: Bytes,
     endpoint: Endpoint,
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use libp2p_test_muxer::test_muxer;
+
+    test_muxer!(empty_payload, MplexConfig::default());
```
[the remark this answers is not in the document]
Comment (dvdplm) [reply muxers/mplex/src/lib.rs]: I see your point for sure. My thinking was that it'd evolve to do more, to start with it'd take a list of tests, like 

```rust
test_muxer!([empty_payload, one_megabyte, ten_megabytes, bidirectional, interrupted_server, rate_limited_server, interrupted_client, rate_limited_client], Config::default());
test_muxer!([empty_payload, one_megabyte, ten_megabytes, bidirectional], special_config);
``` 

Another motive was that the original issue asked for a single method, `test_muxer()` to be called from implementations, as a kind of black-box almost. That doesn't work so well in the presence of different configuration settings for different muxers and so this was a tentative compromise.

That said, I totally agree that as it stands it hardly carries its weight.

File: muxers/test-muxer/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+            $crate::MuxerTester::$test_name($config);
+        }
+    };
+}
+
+pub struct MuxerTester<U, O, E> {
+    _upgrade: PhantomData<U>,
+    _muxer: PhantomData<O>,
+    _error: PhantomData<E>,
+}
```
[the remark this answers is not in the document]
Comment (dvdplm) [reply muxers/test-muxer/src/lib.rs]: Mostly as a way to have the big `where` clause in a single place. As with the macro, I had vague plans to extend this to carry some state further on. It was also a nice enough way to add helpers to init loggers, setup futures etc.

File: muxers/mplex/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (tomaka) [reply muxers/mplex/src/lib.rs]: I don't think we need multiple configuration settings. All the muxers should be tested the same way. It's the `test_muxer()` itself that would dispatch to the different tests.

[the remark this answers is not in the document]
Comment (dvdplm) [reply muxers/mplex/src/lib.rs]: > All the muxers should be tested the same way. 

This is problematic I think. For some configuration options we could make an effort to normalize settings between impls: mplex has `max_substreams` vs yamux `max_num_streams`, mplex has `max_buffer_size` vs yamux `max_buffer_len`. 
Other things are deeply linked to the specific impl though. Take for instance the difference in behaviour between [`OnReceive`](https://github.com/paritytech/yamux/blob/master/src/lib.rs#L42) and [`OnRead`](https://github.com/paritytech/yamux/blob/master/src/lib.rs#L56).

For these reasons I think it makes sense to be able to run tests using different configs.

> It's the test_muxer() itself that would dispatch to the different tests.

I started out like this and it became annoying to work on the tests dispatched this way; for instance to filter out some tests I have to comment out code. I also don't like not seeing the tests printed out one by one as they run, but maybe that's just me.

Why is it important to have a single function call?

File: muxers/test-muxer/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+
+#[macro_export]
+macro_rules! test_muxer {
+    ($test_name: ident, $config: expr) => {
+        #[test]
+        fn $test_name() {
+            $crate::MuxerTester::$test_name($config);
+        }
+    };
+}
```
[the remark this answers is not in the document]
Comment (dvdplm) [reply muxers/test-muxer/src/lib.rs]: Removed the macro.

File: muxers/mplex/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (tomaka) [reply muxers/mplex/src/lib.rs]: I guess it makes sense to have different configurations when you're testing corner cases (which you definitely want to test).
However it would be great if it was as generic as possible, and not rely too much on details that are *too* specific.

**Comment by romanb** [comment]
I guess this is too out of date by now.
